### DISCUSSION TITLE
mixed usage of AD and non-AD materials and kernels

### DISCUSSION THREAD

**Discussion by Bearunner** [idaholab/moose#26830] [Q&A Modules: Porous Flow] [answered]
Dear all,

I add one finite element AD kernel and use it with other non-AD kenels in Porous Flow. 

I need to use the AD and non-AD property 'PorousFlow_fluid_phase_density_nodal' in these kernels. When I run my input, I get the following message:

> The requested AD material property 'PorousFlow_fluid_phase_density_nodal' of type 'std::vector<double>'
> is already retrieved or declared as a non-AD property of type 'std::vector<double>'.

It seems it is not allowed to use the AD and non-AD property 'PorousFlow_fluid_phase_density_nodal' in one input file.

Is there a solution for this? Are the AD materials and kernels only used in finite volume parts of Porous flow modules in its current status?

Thanks.

B

**Comment by GiudGiud** [comment]
Hello

There is a material to create the converted AD or non AD version of a property.

It s the MaterialADConvertor and it s counterpart for the other direction of conversion

https://mooseframework.inl.gov/source/materials/MaterialADConverter.html

Guillaume

**Reply by Bearunner** [reply]
thanks for your tips.

I add the following contents in my input. 

 ```
  [convert_to_AD]
    type = MaterialADConverter
  	reg_props_in = 'PorousFlow_fluid_phase_density_nodal'
  	ad_props_out = 'AD_PorousFlow_fluid_phase_density_nodal'
  []
```

And I change 

   **_fluid_density_node(getADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_density_nodal")),**
into
    **_fluid_density_node(getADMaterialProperty<std::vector<Real>>("AD_PorousFlow_fluid_phase_density_nodal")),**

in my own AD kernel.

When I run my input, I get the following error: 
```
***** ERROR ***
The following error occurred in the object "PorousFlow_density_nodal_all", of type "PorousFlowJoiner".

The declared non-AD material property 'PorousFlow_fluid_phase_density_nodal' of type 'std::vector<double>'
is already retrieved or declared as a non-AD property of type 'double'.
```

I want to use MaterialProperty ("PorousFlow_fluid_phase_density_nodal") in non-AD Kernel and  ADMaterialProperty("AD PorousFlow_fluid_phase_density_nodal") in AD Kernel.

**It seems that every property can only be one type of AD or non-AD, not both types in one input file.**

Could you please help with this? Maybe adding a new AD material is an easier way?

Thanks again.

B

**Reply by GiudGiud** [reply]
The error is complaining about a difference between a vector and a double (single floating point number) type

So this is not about AD conversion anymore

you need to re examine the types for the creation and consumption of the material properties. And make sure they match with your code.

**Reply by cpgr** [reply]
I think that both the AD and non-AD material property have the same name, so this won't work? I'm guessing that you have both of these
```
[fluid]
  type = PorousFlowSingleComponentFluid
[]
[adfluid]
  type = ADPorousFlowSingleComponentFluid
[]
```
but they both make a material with the same name, and then the other objects don't know which to access?

Unfortunately, you might have to modify the code to do this. Perhaps we should add an `ad_` to the front of the material property name when the ad version is used? @GiudGiud - do you think this is the way to go? It wouldn't be hard to do. I hadn't considered someone trying to use both AD and non-AD version of a material at the same time.

**Reply by GiudGiud** [reply]
I dont think the prefix is necessary unless
- you start doing this in normal operations 
- you do not let users specify the names they want for material properties.

The [MaterialADConverter](https://mooseframework.inl.gov/source/materials/MaterialADConverter.html) lets you use one of the two Materials and manually add the prefix there. Then in AD objects you specify the prefix

From the error message mentioning 'double' vs `std::vector` I dont think this is just AD vs non-AD anymore

**Reply by Bearunner** [reply]
Thanks for your tips @GiudGiud and @cpgr .

1. To avoid both the AD and non-AD material properties having the same name, I added a new property named 'for_AD_PorousFlow_fluid_phase_density_nodal' totally same to the 'PorousFlow_fluid_phase_density_nodal' in Material: [PorousFlowActionBase.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/materials/PorousFlowFluidPropertiesBase.C)

    _density(_compute_rho_mu
                 ? (_nodal_material ? &this->template declareGenericProperty<Real, is_ad>(
                                          "PorousFlow_fluid_phase_density_nodal" + _phase)
                                    : &this->template declareGenericProperty<Real, is_ad>(
                                          "PorousFlow_fluid_phase_density_qp" + _phase))
                 : nullptr),
   _for_AD_density(_compute_rho_mu
                ? (_nodal_material ? &this->template declareGenericProperty<Real, is_ad>(
                                          "for_AD_PorousFlow_fluid_phase_density_nodal" + _phase)
                                   : &this->template declareGenericProperty<Real, is_ad>(
                                          "for_AD_PorousFlow_fluid_phase_density_qp" + _phase))
                : nullptr),

in Material: [PorousFlowSingleComponentFluid.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C)

    (*_density)[_qp] = _fp.rho_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                        _temperature[_qp] + _t_c2k);

    (*_for_AD_density)[_qp] = _fp.rho_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                                                            _temperature[_qp] + _t_c2k);

in Action: [PorousFlowAddMaterialJoiner.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C)

            if (params.get<bool>("compute_density_and_viscosity"))
            {
              if (at_nodes)
              {
                addJoiner(at_nodes,
                          is_ad,
                          "PorousFlow_fluid_phase_density_nodal",
                          "PorousFlow_density_nodal_all");
                addJoiner(at_nodes,
                           is_ad,
                          "for_AD_PorousFlow_fluid_phase_density_nodal",
                          "for_AD_PorousFlow_density_nodal_all");


in Action: [PorousFlowSinglePhaseBase.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C)

  const bool compute_rho_mu_qp = _deps.dependsOn(_included_objects, "density_qp") ||
                                 _deps.dependsOn(_included_objects, "for_AD_density_qp") ||
                                 _deps.dependsOn(_included_objects, "viscosity_qp");

The above information is the main part of the newly added property named 'for_AD_PorousFlow_fluid_phase_density_nodal', some similar codes are not shown here. 

2. The new property 'for_AD_PorousFlow_fluid_phase_density_nodal' has the same type of 'std::vector<double>' to 'PorousFlow_fluid_phase_density_nodal'   as shown in [PorousFlowMassTimeDerivative.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowMassTimeDerivative.C)

    _fluid_density(_multiply_by_density ? &getMaterialProperty<std::vector<Real>>(
                                              "PorousFlow_fluid_phase_density_nodal")
                                        : nullptr),

I use the MaterialADConverter 

  [convert_to_AD]
    type = MaterialADConverter
    reg_props_in = 'for_AD_PorousFlow_fluid_phase_density_nodal'
    ad_props_out = 'AD_PorousFlow_fluid_phase_density_nodal'
  []

to get the 'AD_PorousFlow_fluid_phase_density_nodal' required in my own AD kernel.

    _ad_fluid_density_node(getADMaterialProperty<std::vector<Real>>("AD_PorousFlow_fluid_phase_density_nodal")),

But I still get the following error information:


*** ERROR ***
The following error occurred in the object "PorousFlow_density_nodal_all", of type "PorousFlowJoiner".

The declared non-AD material property 'PorousFlow_fluid_phase_density_nodal' of type 'std::vector<double>'
is already retrieved or declared as a non-AD property of type 'double'.

**Comment by GiudGiud** [comment, marked as answer]
you declare it as a Real
```
_density(_compute_rho_mu
? (_nodal_material ? &this->template declareGenericProperty<Real, is_ad>(
"PorousFlow_fluid_phase_density_nodal" + _phase)
```
and retrieve it as a vector
```
_ad_fluid_density_node(getADMaterialProperty<std::vector<Real>>("AD_PorousFlow_fluid_phase_density_nodal")),
```

that s the problem

**Reply by Bearunner** [reply]
Thanks a lot.  Yes, you are right! It solves.
